subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/DirectorySensitivity.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.fingerprint;
+
+/**
+ * Specifies how a fingerprinter should handle directories that are found in a filecollection.
+ */
+public enum DirectorySensitivity {
+    // Fingerprint empty directories
+    FINGERPRINT_DIRECTORIES,
+    // Ignore empty directories
+    IGNORE_DIRECTORIES,
+    // Not important (i.e. the fingerprinter does not handle empty directories one way or the other)
+    NONE
```

Comment:
I suspect that if we removed `NONE` in favor of `DEFAULT`, then we could do away with the whole `DirectorySensitiveNormalizer` concept, too.

Reply:
Actually, thinking through this, and with the inclusion of absolute path fingerprinting in the directory normalization, this gets a lot simpler.  There are really only two types of fingerprinters now:

- Those that can be configured to either ignore or fingerprint directories (AbsolutePath, RelativePath, NameOnly) and default to fingerprinting directories.
- Those that do not fingerprint directories at all (Classpath, CompileClasspath, IgnoredPath) and cannot be configured.

`DEFAULT` feels a little squirrelly, because the "default" behavior is different between these two classes (the default for the first class is to fingerprint directories, the default, and only mode, for the second class is to exclude directories) but it does provide the nice advantage that everything is just "whatever the default is" unless you set it explicitly to ignore directories.

I've changed this to `DEFAULT` and `IGNORE_DIRECTORIES`.  I don't think we can get away without `DirectorySensitiveNormalizer` unless we produce an error if the user says something like  `ignoreDirectories().withNormalizer(ClasspathNormalizer)`.  This would search the fingerprinter registry for a fingerprinter registered with the `ClasspathNormalizer` and `IGNORE_DIRECTORIES` which won't exist.  It doesn't feel right to me to error in such a case, since ignoring directories is "valid" behavior for this class of normalizer.  To recognize that ignoring directories _is_ the default behavior and make sure that we search for a fingerprinter registered with `DEFAULT` means we would have to be able to identify which normalizers fall into this category (which would require something like `DirectorySensitiveNormalizer`).

I think the only way to get rid of `DirectorySensitiveNormalizer` would be to register a fingerprinter for both `DEFAULT` and `IGNORE_DIRECTORIES` for those where the default behavior is to ignore or to build some sort of knowledge of this distinction into the registry lookup.  These both feel worse to me.

I've changed this to `DEFAULT` vs `IGNORE_DIRECTORIES` but left `DirectorySensitiveNormalizer` in place, unless you have a better idea that I'm not thinking of.